File: analysis/common_func/rt_add_info_center.py
# -------------------------------------------------------------------------
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This file is part of the MindStudio project.
#
# MindStudio is licensed under Mulan PSL v2.
# You can use this software according to the terms and conditions of the Mulan PSL v2.
# You may obtain a copy of Mulan PSL v2 at:
#
#    http://license.coscl.org.cn/MulanPSL2
#
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
# EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
# MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
# See the Mulan PSL v2 for more details.
# -------------------------------------------------------------------------
import logging
import os
from typing import Optional, Dict, Tuple

from common_func.constant import Constant
from common_func.singleton import singleton
from common_func.path_manager import PathManager
from common_func.db_name_constant import DBNameConstant
from msmodel.add_info.runtime_op_info_model import RuntimeOpInfoViewModel
from msmodel.compact_info.capture_stream_info_model import CaptureStreamInfoViewModel
from profiling_bean.db_dto.runtime_op_info_dto import RuntimeOpInfoDto


class CaptureStatus:
    START = 0
    END = 1


@singleton
class RTAddInfoCenter:

    def __init__(self: any, project_path: str) -> None:
        self._op_info_dict = {}
        self._capture_info_list = []
        if os.path.exists(PathManager.get_db_path(project_path, DBNameConstant.DB_RTS_TRACK)):
            self.load_runtime_op_info_data(project_path)
        if os.path.exists(PathManager.get_db_path(project_path, DBNameConstant.DB_STREAM_INFO)):
            self.load_capture_stream_info_data(project_path)
        self._capture_info_time_range_dict = self.build_capture_info_time_range_dict()

    def load_runtime_op_info_data(self: any, project_path: str) -> None:
        """
        load runtime op info data
        """
        try:
            with RuntimeOpInfoViewModel(project_path) as _model:
                self._op_info_dict = _model.get_runtime_op_info_data()
        except Exception as e:
            logging.error(f"Failed to load runtime op info data: {e}")
            self._op_info_dict = []

    def load_capture_stream_info_data(self: any, project_path: str) -> None:
        """
        load capture stream info data
        """
        try:
            with CaptureStreamInfoViewModel(project_path) as _model:
                self._capture_info_list = _model.get_capture_stream_info_data()
        except Exception as e:
            logging.error(f"Failed to load capture stream info data: {e}")
            self._capture_info_list = []

    def build_capture_info_time_range_dict(self) -> Dict[Tuple[int, int, int], Tuple[int, int, int]]:
        capture_info_time_range_dict = {}
        for capture_info in self._capture_info_list:
            key = (capture_info.device_id, capture_info.stream_id, capture_info.batch_id)

            current_range = capture_info_time_range_dict.get(key, (
                0,
                float("inf"),
                capture_info.model_id
            ))
            if capture_info.capture_status == CaptureStatus.START:
                capture_info_time_range_dict[key] = (
                    capture_info.timestamp,
                    current_range[1],
                    capture_info.model_id
                )
            elif capture_info.capture_status == CaptureStatus.END:
                capture_info_time_range_dict[key] = (
                    current_range[0],
                    capture_info.timestamp,
                    capture_info.model_id
                )
        return capture_info_time_range_dict

    def find_matching_model_id(self, device_id: int, stream_id: int, batch_id: int, timestamp: float) -> Optional[int]:
        search_key = (device_id, stream_id, batch_id)
        if search_key not in self._capture_info_time_range_dict:
            return Constant.GE_OP_MODEL_ID
        start_time, end_time, model_id = self._capture_info_time_range_dict[search_key]
        if start_time <= timestamp <= end_time:
            return model_id
        return Constant.GE_OP_MODEL_ID

    def get_op_info_by_id(self: any, device_id: int, stream_id: int, task_id: int, batch_id: int,
                          timestamp: float) -> Tuple[int, RuntimeOpInfoDto]:
        """
        get type hash dict data
        后续需要增加batchId做唯一id关联，batchId应通过capture stream info数据获取
        """
        return self.find_matching_model_id(device_id, stream_id, batch_id, timestamp), self._op_info_dict.get(
            (device_id, stream_id, task_id), RuntimeOpInfoDto()
        )

File: analysis/csrc/domain/services/parser/host/cann/rt_add_info_center.cpp
/* -------------------------------------------------------------------------
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This file is part of the MindStudio project.
 *
 * MindStudio is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *
 *    http://license.coscl.org.cn/MulanPSL2
 *
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
 * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 * -------------------------------------------------------------------------
*/

#include "analysis/csrc/domain/services/parser/host/cann/rt_add_info_center.h"

#include "analysis/csrc/infrastructure/db/include/database.h"
#include "analysis/csrc/infrastructure/db/include/db_runner.h"
#include "analysis/csrc/infrastructure/dfx/error_code.h"
#include "analysis/csrc/domain/services/parser/host/cann/hash_data.h"

namespace Analysis {
namespace Domain {
namespace Host {
namespace Cann {

using namespace Analysis::Utils;
using namespace Analysis::Infra;

void Cann::RTAddInfoCenter::Load(const std::string &path)
{
    LoadDB(path);
    LoadCaptureInfoDB(path);
    BuildCaptureInfoTimeRange();
}

RuntimeOpInfo RTAddInfoCenter::Get(uint16_t deviceId, uint32_t streamId, uint16_t taskId)
{
    std::string key = Utils::Join("_", deviceId, streamId, taskId);
    auto it = runtimeOpInfoData_.find(key);
    if (it != runtimeOpInfoData_.end()) {
        return it->second;
    }
    return RuntimeOpInfo();
}

uint64_t RTAddInfoCenter::GetModelId(uint16_t deviceId, uint32_t streamId, uint16_t batchId, uint64_t timestamp)
{
    constexpr uint32_t kDefaultModelId = DEFAULT_MODEL_ID;

    CaptureKey key{deviceId, streamId, batchId};
    auto it = captureInfoTimeRangeDict_.find(key);
    if (it == captureInfoTimeRangeDict_.end())
        return kDefaultModelId;

    const auto& timeRange = it->second;
    uint64_t start = std::get<0>(timeRange);
    uint64_t end   = std::get<1>(timeRange);
    uint32_t model = std::get<2>(timeRange);

    return (timestamp >= start && timestamp <= end) ? model : kDefaultModelId;
}

void RTAddInfoCenter::BuildCaptureInfoTimeRange()
{
    const uint64_t kInf = std::numeric_limits<uint64_t>::max();

    for (const auto& info : captureStreamInfoData_) {
        CaptureKey key{info.deviceId, info.streamId, info.batchId};

        auto it = captureInfoTimeRangeDict_.find(key);
        if (it == captureInfoTimeRangeDict_.end()) {
            it = captureInfoTimeRangeDict_.emplace(key, TimeRangeInfo{0, kInf, info.modelId}).first;
        }
        auto& timeRange = it->second;

        if (info.captureStatus == CAPTURE_STATUS_START) {
            std::get<0>(timeRange) = info.timeStamp;   // startTime
            std::get<2>(timeRange) = info.modelId;     // modelId
        } else if (info.captureStatus == CAPTURE_STATUS_END) {
            std::get<1>(timeRange) = info.timeStamp;   // endTime
            std::get<2>(timeRange) = info.modelId;
        }
    }
}

void RTAddInfoCenter::LoadCaptureInfoDB(const std::string &path)
{
    StreamInfoDB streamInfoDB;
    std::string hostDbDirectory = Utils::File::PathJoin({path, streamInfoDB.GetDBName()});
    DBRunner dbRunner(hostDbDirectory);
    if (!File::Exist(hostDbDirectory) || !dbRunner.CheckTableExists("CaptureStreamInfo")) {
        return;
    }
    std::string sql{"SELECT model_id, timestamp, stream_id, original_stream_id, device_id, batch_id, capture_status "
                    "FROM CaptureStreamInfo"};
    using DataFormat = std::vector<std::tuple<uint64_t, uint64_t, uint32_t, uint16_t, uint16_t, uint16_t, uint16_t>>;
    DataFormat result;
    if (!dbRunner.QueryData(sql, result)) {
        ERROR("Query capture stream info data failed, db path is %.", hostDbDirectory);
        return;
    }

    for (const auto& row : result) {
        uint16_t deviceId;
        uint16_t originalStreamId;
        uint16_t batchId;
        uint16_t captureStatus;
        uint32_t streamId;
        uint64_t modelId;
        uint64_t timeStamp;
        std::tie(modelId, timeStamp, streamId, originalStreamId, deviceId, batchId, captureStatus) = row;
        CaptureStreamInfo info{modelId, timeStamp, streamId, originalStreamId, deviceId, batchId, captureStatus};
        captureStreamInfoData_.push_back(info);
    }
}


void RTAddInfoCenter::LoadDB(const std::string &path)
{
    RtsTrackDB rtsTrackDb;
    std::string hostDbDirectory = Utils::File::PathJoin({path, rtsTrackDb.GetDBName()});
    DBRunner dbRunner(hostDbDirectory);
    if (!File::Exist(hostDbDirectory) || !dbRunner.CheckTableExists("RuntimeOpInfo")) {
        return;
    }

    std::string sql{"SELECT device_id, model_id, stream_id, task_id, op_name, task_type, op_type, "
                    "block_num, mix_block_num, op_flag, is_dynamic, tensor_num, input_formats, "
                    "input_data_types, input_shapes, output_formats, output_data_types, output_shapes "
                    "FROM RuntimeOpInfo"};
    using dataFormat = std::vector<std::tuple<uint16_t, uint64_t, uint32_t, uint16_t, std::string, std::string,
                                              std::string, uint16_t, uint16_t, uint16_t, std::string, uint16_t,
                                              std::string, std::string, std::string, std::string, std::string,
                                              std::string>>;
    dataFormat result;
    if (!dbRunner.QueryData(sql, result)) {
        ERROR("Query runtime op info data failed, db path is %.", hostDbDirectory);
        return;
    }

    for (auto row : result) {
        uint16_t deviceId, taskId, blockNum, mixBlockNum, opFlag, tensorNum;
        uint32_t streamId;
        uint64_t modelId, opNameU64, opTypeU64;
        std::string opName, taskType, opType, isDynamic, inputFormats, inputDataTypes, inputShapes,
                    outputFormats, outputDataTypes, outputShapes;
        std::tie(deviceId, modelId, streamId, taskId, opName, taskType, opType, blockNum, mixBlockNum, opFlag,
                 isDynamic, tensorNum, inputFormats, inputDataTypes, inputShapes,
                 outputFormats, outputDataTypes, outputShapes) = row;
        std::string key = Utils::Join("_", deviceId, streamId, taskId);
        if (Utils::StrToU64(opNameU64, opName) != ANALYSIS_OK || Utils::StrToU64(opTypeU64, opType) != ANALYSIS_OK) {
            ERROR("opName hash is %, opType hash is %.", opName, opType);
            continue;
        }
        opName = HashData::GetInstance().Get(opNameU64);
        opType = HashData::GetInstance().Get(opTypeU64);
        RuntimeOpInfo info{deviceId, taskId, blockNum, mixBlockNum, opFlag, tensorNum, streamId, modelId, taskType,
                           opType, opName, isDynamic, inputFormats, inputDataTypes, inputShapes,
                           outputFormats, outputDataTypes, outputShapes};
        runtimeOpInfoData_[key] = info;
    }
}

}  // namespace Cann
}  // namespace Host
}  // namespace Domain
}  // namespace Analysis
